File: app/src/main/java/com/kelsos/mbrc/connection_manager/ConnectionManagerPresenterImpl.java
package com.kelsos.mbrc.connection_manager;

import com.kelsos.mbrc.data.ConnectionSettings;
import com.kelsos.mbrc.presenters.BasePresenter;
import com.kelsos.mbrc.repository.ConnectionRepository;

import java.util.List;

import javax.inject.Inject;

import rx.Observable;


public class ConnectionManagerPresenterImpl extends BasePresenter<ConnectionManagerView>
    implements ConnectionManagerPresenter {

  private ConnectionRepository repository;

  @Inject
  public ConnectionManagerPresenterImpl(ConnectionRepository repository) {
    this.repository = repository;
  }

  @Override
  public void load() {
    checkIfAttached();
    Observable<List<ConnectionSettings>> all = Observable.defer(() -> Observable.just(repository.getAll()));
    Observable<Long> defaultId = Observable.defer(() -> Observable.just(repository.getDefaultId()));

    addSubcription(Observable.zip(defaultId, all, ConnectionModel::create)
        .subscribe(getView()::updateModel, this::onLoadError));
  }

  @Override
  public void setDefault(ConnectionSettings settings) {
    checkIfAttached();
    repository.setDefault(settings);
    getView().defaultChanged();
  }

  @Override
  public void update(ConnectionSettings settings) {
    checkIfAttached();
    repository.update(settings);
    if (settings.getId() == repository.getDefaultId()) {
      getView().defaultChanged();
    }
  }

  @Override
  public void delete(ConnectionSettings settings) {
    checkIfAttached();
    repository.delete(settings);
    if (settings.getId() == repository.getDefaultId()) {
      getView().defaultChanged();
    }
  }

  private void onLoadError(Throwable throwable) {
    checkIfAttached();

  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/ui/connection_manager/ConnectionManagerPresenterImpl.kt
package com.kelsos.mbrc.ui.connection_manager

import com.kelsos.mbrc.mvp.BasePresenter
import com.kelsos.mbrc.networking.connections.ConnectionRepository
import com.kelsos.mbrc.networking.connections.ConnectionSettings
import io.reactivex.Observable
import io.reactivex.functions.BiFunction
import timber.log.Timber
import javax.inject.Inject


class ConnectionManagerPresenterImpl
@Inject
constructor(private val repository: ConnectionRepository) : BasePresenter<ConnectionManagerView>(), ConnectionManagerPresenter {

  override fun load() {
    checkIfAttached()
    val all = Observable.defer { Observable.just(repository.all) }
    val defaultId = Observable.defer { Observable.just(repository.defaultId) }

    addDisposable(Observable.zip<Long,
        List<ConnectionSettings>,
        ConnectionModel>(defaultId, all, BiFunction(::ConnectionModel))
        .subscribe({
          view?.updateModel(it)
        }, {
          this.onLoadError(it)
        }))
  }

  override fun setDefault(settings: ConnectionSettings) {
    checkIfAttached()
    repository.default = settings
    view?.defaultChanged()
    view?.dataUpdated()
  }

  override fun save(settings: ConnectionSettings) {
    checkIfAttached()

    if (settings.id > 0) {
      repository.update(settings)
    } else {
      repository.save(settings)
    }

    if (settings.id == repository.defaultId) {
      view?.defaultChanged()
    }

    view?.dataUpdated()
  }

  override fun delete(settings: ConnectionSettings) {
    checkIfAttached()
    repository.delete(settings)
    if (settings.id == repository.defaultId) {
      view?.defaultChanged()
    }

    view?.dataUpdated()
  }

  private fun onLoadError(throwable: Throwable) {
    checkIfAttached()
    Timber.v(throwable, "Failure")
  }
}
